### PR TITLE
Proper closing of resources in MessageAction

### PR DESCRIPTION
**Pull request by MinnDevelopment**
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

There are several guidelines you should follow in order for your
  Pull Request to be merged.

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

> It is sometimes better to include more changes in a single commit. 
  If you find yourself having an overwhelming amount of commits, you
  can **rebase** your branch.

## Description

Added the following methods

* clearFiles(Consumer)
* clearFiles(BiConsumer)

This should allow users to close any of their provided resources,
additionally clearFiles() will now close any opened FileInputStreams!

I've added a note to the class-level documentation that the execution of MessageAction closes all provided resources.
You cannot send the same resource twice without copying it first!

### CODE REVIEW

File: src/main/java/net/dv8tion/jda/core/requests/restaction/MessageAction.java [hunk 1]
```diff
@@ -49,6 +49,8 @@
  * <br>This can be used to remove existing embeds from a message:
  * <br>{@code message.editMessage("This message had an embed").override(true).queue()}
  *
+ * <p><u>When this RestAction has been executed all provided files will be closed.</u></p>
```
Comment (kantenkugel) [inline src/main/java/net/dv8tion/jda/core/requests/restaction/MessageAction.java]: I would rephrase this to something like: "If you provide Files to this builder but decide not to actually send this message, use clearFiles() to clean up their resources. See description of sendFile(File)"

File: src/main/java/net/dv8tion/jda/core/requests/restaction/MessageAction.java [hunk 2]
```diff
@@ -337,6 +340,7 @@ public MessageAction appendFormat(final String format, final Object... args)
 
     /**
      * Adds the provided {@link java.io.InputStream InputStream} as file data.
+     * <br><u>The stream will be closed upon execution!</u>
```
Comment (kantenkugel) [inline src/main/java/net/dv8tion/jda/core/requests/restaction/MessageAction.java]: When are given input streams actually closed? in the sending code?

File: src/main/java/net/dv8tion/jda/core/requests/restaction/MessageAction.java [hunk 3]
```diff
@@ -436,6 +440,7 @@ public MessageAction addFile(final File file)
 
     /**
      * Adds the provided {@link java.io.File File} as file data.
+     * <br><u>The file stream will be closed upon execution!</u>
```
Comment (kantenkugel) [inline src/main/java/net/dv8tion/jda/core/requests/restaction/MessageAction.java]: Would rephrase this similar to the suggestion of class definition ("this method creates a stream for reading the file, which should be cleaned up via clearFiles()")

File: src/main/java/net/dv8tion/jda/core/requests/restaction/MessageAction.java [hunk 4]
```diff
@@ -480,13 +487,67 @@ public MessageAction addFile(final File file, final String name)
...
+     *
+     * @see    #clearFiles(Consumer)
+     * @see    #clearFiles(BiConsumer)
      */
     @CheckReturnValue
     public MessageAction clearFiles()
     {
         files.clear();
+        clearResources();
+        return this;
```
Comment (kantenkugel) [inline src/main/java/net/dv8tion/jda/core/requests/restaction/MessageAction.java]: Would by default close un-owned streams too (knowing ppl won't do it manually) and add an override with boolean parameter closeUnowned or something. same goes for other clearFiles() methods

File: src/main/java/net/dv8tion/jda/core/requests/restaction/MessageAction.java [hunk 5]
```diff
@@ -517,6 +595,9 @@ protected RequestBody asMultipart()
         }
         if (!isEmpty())
             builder.addFormDataPart("payload_json", getJSON().toString());
+        // clear remaining resources, they will be closed after being sent
+        files.clear();
+        ownedResources.clear();
```
Comment (kantenkugel) [inline src/main/java/net/dv8tion/jda/core/requests/restaction/MessageAction.java]: not sure if this is only called by actual execution of the restaction (and if that will cause the MultipartBody to close all given streams)

File: src/main/java/net/dv8tion/jda/core/requests/restaction/MessageAction.java (hunk 2 above)
Comment (MinnDevelopment) [reply src/main/java/net/dv8tion/jda/core/requests/restaction/MessageAction.java]: When they are read by the RequestBody it closes the stream.

File: src/main/java/net/dv8tion/jda/core/requests/restaction/MessageAction.java (hunk 5 above)
Comment (MinnDevelopment) [reply src/main/java/net/dv8tion/jda/core/requests/restaction/MessageAction.java]: The body does close the stream, it is only called on execution.

File: src/main/java/net/dv8tion/jda/core/requests/restaction/MessageAction.java (hunk 4 above)
Comment (MinnDevelopment) [reply src/main/java/net/dv8tion/jda/core/requests/restaction/MessageAction.java]: I don't want to close streams that were never used

Comment (kantenkugel) [reply src/main/java/net/dv8tion/jda/core/requests/restaction/MessageAction.java]: Yea but that would mean if a user uses channel.sendFile(someInputStream, "name") and never actually activates the restaction (or just calls clearFiles()), they would have to manually close the inputstream on their end. I think it would be nice if we would close any inputstream we get if they are either sent or the MessageAction is "canceled" (clearFiles).

Therefore i also suggested that rewrite of the docs (and probably add a big warning)

Comment (MinnDevelopment) [reply src/main/java/net/dv8tion/jda/core/requests/restaction/MessageAction.java]: If the user wants to close _all_ streams they can do so with `clearFiles(Consumer)`

Comment (kantenkugel) [reply src/main/java/net/dv8tion/jda/core/requests/restaction/MessageAction.java]: oh right... add that to some jdocs tho
